File: flask-server/ai_manager.py
from abc import ABC, abstractmethod
import anthropic
import tiktoken
from openai import OpenAI
from email import message
from multiprocessing.managers import Value

class AIProvider(ABC):
    @abstractmethod
    def get_all_models(self):
        pass

    @abstractmethod
    def get_current_model(self):
        pass

    @abstractmethod
    def set_model(self, model):
        pass

    @abstractmethod
    def get_token_count(self, messages):
        pass

    @abstractmethod
    def get_response(self, chat):
        pass

    @abstractmethod
    def get_response_stream(self, messages, system):
        pass

class DeepseekProvider(AIProvider):
    MODELS = {
        "deepseek-chat": {
            "id": "deepseek-chat",
            "display_name": "Deepseek V3",
            "context_window": 64000,
            "input_token_cost": 0.27,
            "output_token_cost": 1.1
        },
    }

    def __init__(self, model="deepseek-chat"):
        self._client = OpenAI(base_url="https://api.deepseek.com")
        self._model = self.MODELS[model]

    def get_all_models(self):
        return list(self.MODELS.values())

    def get_current_model(self):
        return self._model

    def set_model(self, model):
        if model in self.MODELS:
            self._model = self.MODELS[model]
        else:
            raise ValueError(f"Model {model} not found")

    def get_token_count(self, messages):
        user_messages = [msg["content"][0]["text"] for msg in messages if msg["role"] == "user"]
        assistant_messages = [msg["content"] for msg in messages if msg["role"] == "assistant"]
        enc = tiktoken.get_encoding("o200k_base")

        try:
            user_token_count = sum(len(enc.encode(msg)) for msg in user_messages)
            assistant_token_count = sum(len(enc.encode(msg)) for msg in assistant_messages)

            return {
                "input_tokens": user_token_count,
                "output_tokens": assistant_token_count,
            }
        except Exception as e:
            print(f"Error counting tokens: {str(e)}")
            return {
                "input_tokens": 0,
                "output_tokens": 0
            }

    def get_response(self, chat):
        response = self._client.chat.completions.create(
            model=self._model["id"],
            messages=chat,
            stream=False
        )

        return response.choices[0].message.content

    def get_response_stream(self, messages, system):
        stream = self._client.chat.completions.create(
            model=self._model["id"],
            messages=messages,
            stream=True
        )

        for chunk in stream:
            yield chunk.choices[0].delta.content


class ClaudeProvider(AIProvider):
    MODELS = {
        "claude-3-5-sonnet-20241022": {
            "id": "claude-3-5-sonnet-20241022",
            "display_name": "Claude 3.5 Sonnet (New)",
            "context_window": 200000,
            "input_token_cost": 3,
            "output_token_cost": 15
        },
        "claude-3-5-haiku-20241022": {
            "id": "claude-3-5-haiku-20241022",
            "display_name": "Claude 3.5 Haiku",
            "context_window": 200000,
            "input_token_cost": 1,
            "output_token_cost": 5
        }
    }

    def __init__(self, model="claude-3-5-sonnet-20241022"):
        self._client = anthropic.Anthropic()
        self._model = self.MODELS[model]

    def get_all_models(self):
        return list(self.MODELS.values())

    def get_current_model(self):
        return self._model

    def set_model(self, model):
        if model in self.MODELS:
            self._model = self.MODELS[model]
        else:
            raise ValueError(f"Model {model} not found")

    def get_token_count(self, messages):
        user_messages = [msg for msg in messages if msg["role"] == "user"]
        assistant_messages = [msg for msg in messages if msg["role"] == "assistant"]

        try:
            user_token_count = self._client.messages.count_tokens(
                model=self._model["id"],
                messages=user_messages
            ).input_tokens if user_messages else 0

            # Count tokens for assistant messages
            assistant_token_count = self._client.messages.count_tokens(
                model=self._model["id"],
                messages=assistant_messages
            ).input_tokens if assistant_messages else 0

            return {
                "input_tokens": user_token_count,
                "output_tokens": assistant_token_count,
            }
        except Exception as e:
            print(f"Error counting tokens: {str(e)}")
            return {
                "input_tokens": 0,
                "output_tokens": 0
            }

    def get_response(self, chat):
        response = self._client.messages.create(
            model=self._model["id"],
            max_tokens=1024,
            messages=chat
        )

        return response

    def get_response_stream(self, messages, system):
        with self._client.messages.stream(
            max_tokens=1024,
            messages=messages,
            system=system,
            model=self._model["id"]
        ) as stream:
            for text in stream.text_stream:
                yield text

class AIManager:
    PROVIDERS = {
        "claude": {
            "name": "Anthropic",
            "class": ClaudeProvider
        },
        "deepseek": {
            "name": "Deepseek",
            "class": DeepseekProvider
        }
    }

    def __init__(self, provider="deepseek", model=None):
        if provider not in self.PROVIDERS:
            raise ValueError(f"Provider {provider} not supported")

        self.current_provider = provider
        self.provider_instance = self.PROVIDERS[provider]["class"](model=model) if model else self.PROVIDERS[provider]["class"]()

    def get_all_providers(self):
        return {k: v["name"] for k, v in self.PROVIDERS.items()}

    def get_current_provider(self):
        return {
            "id": self.current_provider,
            "name": self.PROVIDERS[self.current_provider]["name"]
        }

    def set_provider(self, provider, model=None):
        if not provider:
            raise ValueError("Provider ID cannot be empty")

        if provider not in self.PROVIDERS:
            raise ValueError(f"Provider '{provider}' not supported. Available providers: {list(self.PROVIDERS.keys())}")

        try:
            self.current_provider = provider
            self.provider_instance =  self.PROVIDERS[provider]["class"]()
            return self.get_current_provider()
        except Exception as e:
            raise Exception(f"Failed to initialize provider '{provider}': {str(e)}")

    def get_all_models(self):
        return self.provider_instance.get_all_models()

    def get_current_model(self):
        return self.provider_instance.get_current_model()

    def set_model(self, model):
        return self.provider_instance.set_model(model)

    def get_token_count(self, messages):
        return self.provider_instance.get_token_count(messages)

    def get_response(self, chat):
        return self.provider_instance.get_response(chat)

    def get_response_stream(self, messages, system):
        return self.provider_instance.get_response_stream(messages, system)

File: src/services/AiManager.js
import { OpenAI } from "openai";
import Anthropic from "@anthropic-ai/sdk";
import { encoding_for_model } from "tiktoken";
import ollama from "ollama";

// Abstract class equivalent
class AIProvider {
  get_all_models() {
    throw new Error("Method 'get_all_models()' must be implemented");
  }

  get_current_model() {
    throw new Error("Method 'get_current_model()' must be implemented");
  }

  set_model(model) {
    throw new Error("Method 'set_model()' must be implemented");
  }

  get_token_count(messages) {
    throw new Error("Method 'get_token_count()' must be implemented");
  }

  get_response(chat) {
    throw new Error("Method 'get_response()' must be implemented");
  }

  get_response_stream(messages, system) {
    throw new Error("Method 'get_response_stream()' must be implemented");
  }
}

class DeepseekProvider extends AIProvider {
  static MODELS = {
    "deepseek-chat": {
      id: "deepseek-chat",
      display_name: "Deepseek V3",
      context_window: 64000,
      input_token_cost: 0.27,
      output_token_cost: 1.1,
    },
  };

  constructor(model = "deepseek-chat") {
    super();
    this._client = new OpenAI({
      baseURL: "https://api.deepseek.com",
    });
    this._model = DeepseekProvider.MODELS[model];
  }

  get_all_models() {
    return Object.values(DeepseekProvider.MODELS);
  }

  get_current_model() {
    return this._model;
  }

  set_model(model) {
    if (DeepseekProvider.MODELS[model]) {
      this._model = DeepseekProvider.MODELS[model];
    } else {
      throw new Error(`Model ${model} not found`);
    }
  }

  get_token_count(messages) {
    try {
      const enc = encoding_for_model("gpt-4o");
      let userTokenCount = 0;
      let assistantTokenCount = 0;

      messages.forEach((msg) => {
        const content = msg.content;
        let text = "";

        if (typeof content === "string") {
          text = content;
        } else if (Array.isArray(content)) {
          text = content
            .map((part) => (part.type === "text" ? part.text : ""))
            .join("");
        }

        if (msg.role === "user") {
          userTokenCount += enc.encode(text).length;
        } else if (msg.role === "assistant") {
          assistantTokenCount += enc.encode(text).length;
        }
      });

      return {
        input_tokens: userTokenCount,
        output_tokens: assistantTokenCount,
      };
    } catch (error) {
      console.error(`Error counting tokens: ${error.message}`);
      return {
        input_tokens: 0,
        output_tokens: 0,
      };
    }
  }

  async get_response(chat) {
    const response = await this._client.chat.completions.create({
      model: this._model.id,
      messages: chat,
      stream: false,
    });
    return response.choices[0].message.content;
  }

  async *get_response_stream(messages, system) {
    const allMessages = [...messages];
    if (system) {
      allMessages.unshift({ role: "system", content: system });
    }

    const stream = await this._client.chat.completions.create({
      model: this._model.id,
      messages: allMessages,
      stream: true,
    });

    for await (const chunk of stream) {
      if (chunk.choices[0].delta.content) {
        yield chunk.choices[0].delta.content;
      }
    }
  }
}

class ClaudeProvider extends AIProvider {
  static MODELS = {
    "claude-3-7-sonnet-20250219": {
      id: "claude-3-7-sonnet-20250219",
      display_name: "Claude 3.7 Sonnet",
      context_window: 200000,
      input_token_cost: 3,
      output_token_cost: 15,
    },
    "claude-3-5-haiku-20241022": {
      id: "claude-3-5-haiku-20241022",
      display_name: "Claude 3.5 Haiku",
      context_window: 200000,
      input_token_cost: 1,
      output_token_cost: 5,
    },
  };

  constructor(model = "claude-3-7-sonnet-20250219") {
    super();
    this._client = new Anthropic();
    this._model = ClaudeProvider.MODELS[model];
  }

  get_all_models() {
    return Object.values(ClaudeProvider.MODELS);
  }

  get_current_model() {
    return this._model;
  }

  set_model(model) {
    if (ClaudeProvider.MODELS[model]) {
      this._model = ClaudeProvider.MODELS[model];
    } else {
      throw new Error(`Model ${model} not found`);
    }
  }

  get_token_count(messages) {
    try {
      let userTokenCount = 0;
      let assistantTokenCount = 0;

      messages.forEach((msg) => {
        const text =
          typeof msg.content === "string"
            ? msg.content
            : msg.content.map((part) => part.text).join("");

        const count = this._client.countTokens(text);

        if (msg.role === "user") {
          userTokenCount += count;
        } else if (msg.role === "assistant") {
          assistantTokenCount += count;
        }
      });

      return {
        input_tokens: userTokenCount,
        output_tokens: assistantTokenCount,
      };
    } catch (error) {
      console.error(`Error counting tokens: ${error.message}`);
      return {
        input_tokens: 0,
        output_tokens: 0,
      };
    }
  }

  async get_response(chat) {
    try {
      // Normalize messages to string format
      const normalizedMessages = this.normalizeMessages(chat);

      const response = await ollama.chat({
        model: this._model.id,
        messages: normalizedMessages,
        stream: false,
      });

      return response.message.content;
    } catch (error) {
      console.error("Ollama API error:", error);
      throw new Error(`Ollama error: ${error.message}`);
    }
  }

  async *get_response_stream(messages, system) {
    try {
      // Normalize messages to string format
      const normalizedMessages = this.normalizeMessages(messages);

      // Add system prompt as a separate message
      if (system) {
        normalizedMessages.unshift({
          role: "system",
          content: system,
        });
      }

      const response = await ollama.chat({
        model: this._model.id,
        messages: normalizedMessages,
        stream: true,
      });

      for await (const chunk of response) {
        if (chunk.message?.content) {
          yield chunk.message.content;
        }
      }
    } catch (error) {
      console.error("Ollama streaming error:", error);
      throw new Error(`Ollama streaming error: ${error.message}`);
    }
  }

  // Convert Claude-style message format to simple string format
  normalizeMessages(messages) {
    return messages.map((msg) => {
      let content = "";

      if (typeof msg.content === "string") {
        content = msg.content;
      } else if (Array.isArray(msg.content)) {
        // Concatenate all text parts
        content = msg.content
          .filter((part) => part.type === "text")
          .map((part) => part.text)
          .join("\n");
      }

      return {
        role: msg.role,
        content: content,
      };
    });
  }

  #extractSystemMessage(messages) {
    let system = "";
    const filteredMessages = [];

    messages.forEach((msg) => {
      if (msg.role === "system") {
        system +=
          typeof msg.content === "string"
            ? msg.content + "\n"
            : msg.content.map((part) => part.text).join("\n");
      } else {
        filteredMessages.push(msg);
      }
    });

    return { system, filteredMessages };
  }
}

class OllamaProvider extends AIProvider {
  static MODELS = {
    "gpt-oss:20b": {
      id: "gpt-oss:20b",
      display_name: "GPT OSS",
      context_window: 128000,
      input_token_cost: 0,
      output_token_cost: 0,
    },
    "qwen2.5-coder:7b": {
      id: "qwen2.5-coder:7b",
      display_name: "Qwen 2.5 Coder 7b",
      context_window: 128000,
      input_token_cost: 0,
      output_token_cost: 0,
    },
  };

  constructor(model = "gpt-oss:20b") {
    super();
    this.set_model(model); // Use set_model to initialize
  }

  get_all_models() {
    return Object.values(OllamaProvider.MODELS);
  }

  get_current_model() {
    return this._model;
  }

  set_model(model) {
    if (OllamaProvider.MODELS[model]) {
      this._model = OllamaProvider.MODELS[model];
    } else {
      OllamaProvider.MODELS[model] = {
        id: model,
        display_name: model,
        context_window: 4096,
        input_token_cost: 0,
        output_token_cost: 0,
      };
      this._model = OllamaProvider.MODELS[model];
    }
  }

  get_token_count(messages) {
    // Simple approximation (4 characters ≈ 1 token)
    let totalChars = 0;

    messages.forEach((msg) => {
      if (typeof msg.content === "string") {
        totalChars += msg.content.length;
      } else if (Array.isArray(msg.content)) {
        msg.content.forEach((item) => {
          if (item.text) totalChars += item.text.length;
        });
      }
    });

    return {
      input_tokens: Math.floor(totalChars / 4),
      output_tokens: 0,
    };
  }

  async get_response(chat) {
    try {
      // Normalize messages to string format
      const normalizedMessages = this.normalizeMessages(chat);

      const response = await ollama.chat({
        model: this._model.id,
        messages: normalizedMessages,
        stream: false,
      });

      return response.message.content;
    } catch (error) {
      console.error("Ollama API error:", error);
      throw new Error(`Ollama error: ${error.message}`);
    }
  }

  async *get_response_stream(messages, system) {
    try {
      // Normalize messages to string format
      const normalizedMessages = this.normalizeMessages(messages);

      // Add system prompt as a separate message
      if (system) {
        normalizedMessages.unshift({
          role: "system",
          content: system,
        });
      }

      const response = await ollama.chat({
        model: this._model.id,
        messages: normalizedMessages,
        stream: true,
      });

      for await (const chunk of response) {
        if (chunk.message?.content) {
          yield chunk.message.content;
        }
      }
    } catch (error) {
      console.error("Ollama streaming error:", error);
      throw new Error(`Ollama streaming error: ${error.message}`);
    }
  }

  // Convert Claude-style message format to simple string format
  normalizeMessages(messages) {
    return messages.map((msg) => {
      let content = "";

      if (typeof msg.content === "string") {
        content = msg.content;
      } else if (Array.isArray(msg.content)) {
        // Concatenate all text parts
        content = msg.content
          .filter((part) => part.type === "text")
          .map((part) => part.text)
          .join("\n");
      }

      return {
        role: msg.role,
        content: content,
      };
    });
  }
}

class AIManager {
  static PROVIDERS = {
    claude: {
      name: "Anthropic",
      class: ClaudeProvider,
    },
    deepseek: {
      name: "Deepseek",
      class: DeepseekProvider,
    },
    ollama: {
      name: "Ollama",
      class: OllamaProvider,
    },
  };

  constructor(provider = "ollama", model = "gemma3n:e4b") {
    if (!AIManager.PROVIDERS[provider]) {
      throw new Error(`Provider ${provider} not supported`);
    }

    this.current_provider = provider;
    const ProviderClass = AIManager.PROVIDERS[provider].class;
    this.provider_instance = model
      ? new ProviderClass(model)
      : new ProviderClass();
  }

  get_all_providers() {
    return Object.fromEntries(
      Object.entries(AIManager.PROVIDERS).map(([id, config]) => [
        id,
        config.name,
      ]),
    );
  }

  get_current_provider() {
    return {
      id: this.current_provider,
      name: AIManager.PROVIDERS[this.current_provider].name,
    };
  }

  set_provider(provider, model = null) {
    if (!provider) {
      throw new Error("Provider ID cannot be empty");
    }

    if (!AIManager.PROVIDERS[provider]) {
      const available = Object.keys(AIManager.PROVIDERS).join(", ");
      throw new Error(
        `Provider '${provider}' not supported. Available providers: ${available}`,
      );
    }

    try {
      this.current_provider = provider;
      const ProviderClass = AIManager.PROVIDERS[provider].class;
      this.provider_instance = model
        ? new ProviderClass(model)
        : new ProviderClass();

      return this.get_current_provider();
    } catch (error) {
      throw new Error(
        `Failed to initialize provider '${provider}': ${error.message}`,
      );
    }
  }

  get_all_models() {
    return this.provider_instance.get_all_models();
  }

  get_current_model() {
    return this.provider_instance.get_current_model();
  }

  set_model(model) {
    return this.provider_instance.set_model(model);
  }

  get_token_count(messages) {
    return this.provider_instance.get_token_count(messages);
  }

  async get_response(chat) {
    return this.provider_instance.get_response(chat);
  }

  async *get_response_stream(messages, system) {
    for await (const chunk of this.provider_instance.get_response_stream(
      messages,
      system,
    )) {
      yield chunk;
    }
  }
}

export { AIManager, DeepseekProvider, ClaudeProvider, OllamaProvider };
